File: python/src/scientific_workflow/dependencies.py
"""Typed access to immutable results from Workflow's declared dependencies.

Snapshot selection performs no scientific I/O and never broadens runtime scope.
The core module has no NumPy dependency. See api.md for the complete contract.
"""
from __future__ import annotations

import copy
import json
import os
from dataclasses import dataclass
from pathlib import Path
from typing import Generic, Iterator, TypeVar


class DependencyError(ValueError):
    """Invalid dependency snapshot or selection."""


class MissingDependencyError(DependencyError):
    """No result satisfies a selection requiring exactly one."""


class AmbiguousDependencyError(DependencyError):
    """Multiple results satisfy a selection requiring at most one."""

    def __init__(self, selection: str, matches: tuple[str, ...]):
        self.selection, self.matches = selection, matches
        super().__init__(f"ambiguous dependency {selection}: {', '.join(matches)}; select a phase or task")


@dataclass(frozen=True, slots=True)
class RecordingDependency:
    phase: str
    task: str
    execution_unit: str
    member: str
    final_iteration: int
    directory: Path


@dataclass(frozen=True, slots=True)
class ProgramDependency:
    phase: str
    task: str
    directory: Path
    executable: Path
    python_script: Path | None


@dataclass(frozen=True, slots=True)
class NpyDependency:
    phase: str
    task: str
    directory: Path


T = TypeVar("T", RecordingDependency, ProgramDependency, NpyDependency)


@dataclass(frozen=True, slots=True)
class Selection(Generic[T]):
    """Immutable intersection of exact selectors, in snapshot order."""
    _entries: tuple[T, ...]
    _filters: tuple[str, ...] = ()

    def _filter(self, field: str, value: str) -> Selection[T]:
        return Selection(tuple(e for e in self._entries if getattr(e, field, None) == value),
                         (*self._filters, f"{field}={value!r}"))

    def in_phase(self, phase: str) -> Selection[T]:
        return self._filter("phase", phase)

    def task(self, identity: str) -> Selection[T]:
        return self._filter("task", identity)

    def execution_unit(self, key: str) -> Selection[T]:
        """Restrict recording results to an execution-unit key."""
        return self._filter("execution_unit", key)

    def member(self, identity: str) -> Selection[T]:
        """Restrict recording results to a member identity."""
        return self._filter("member", identity)

    def one(self) -> T:
        result = self.optional()
        if result is None:
            raise MissingDependencyError(f"no dependency matches {self._filters}")
        return result

    def optional(self) -> T | None:
        if len(self._entries) > 1:
            raise AmbiguousDependencyError(str(self._filters), tuple(
                f"{e.phase}/{e.task}" + (f"/{e.member}" if isinstance(e, RecordingDependency) else "")
                for e in self._entries))
        return self._entries[0] if self._entries else None

    def __iter__(self) -> Iterator[T]:
        return iter(self._entries)

    def iter(self) -> Iterator[T]:
        return iter(self)


def _name(value: object) -> str:
    if not isinstance(value, str) or not value or value.strip() != value:
        raise DependencyError("expected a nonempty identifier without surrounding whitespace")
    return value


def _path(value: object) -> Path:
    if not isinstance(value, str) or not Path(value).is_absolute():
        raise DependencyError(f"expected absolute path, got {value!r}")
    return Path(value)


def _list(value: object) -> list:
    if not isinstance(value, list):
        raise DependencyError("expected an array in dependency snapshot")
    return value


def _unique(value: str, seen: set[str]) -> None:
    if value in seen:
        raise DependencyError(f"duplicate dependency identity {value!r}")
    seen.add(value)


class Dependencies:
    """Validated dependency snapshot, including preserved unknown workload kinds."""

    def __init__(self, snapshot: object):
        self._raw = copy.deepcopy(snapshot)
        recordings, programs, batches = [], [], []
        try:
            phases_seen = set()
            for phase in _list(self._raw):
                phase_name = _name(phase["phase"])
                _unique(phase_name, phases_seen)
                tasks_seen = set()
                for task in _list(phase["tasks"]):
                    identity = _name(task["identity"])
                    _unique(identity, tasks_seen)
                    directory = _path(task["output_directory"])
                    workload = task["workload"]
                    kind = _name(workload["kind"])
                    if kind == "execution_unit":
                        key = _name(workload["execution_unit"])
                        members = _list(workload["members"])
                        if not members:
                            raise DependencyError("execution unit has no members")
                        members_seen = set()
                        for member in members:
                            name = _name(member["identity"])
                            _unique(name, members_seen)
                            iteration = member["final_iteration"]
                            if type(iteration) is not int or not 0 <= iteration <= 2**64 - 1:
                                raise DependencyError("final_iteration must be a u64")
                            recordings.append(RecordingDependency(phase_name, identity, key, name, iteration, _path(member["output_directory"])))
                    elif kind in ("program", "python"):
                        script = workload.get("python_script")
                        if kind == "python" and script is None:
                            raise DependencyError("python workload requires python_script")
                        programs.append(ProgramDependency(phase_name, identity, directory / "artifacts", _path(workload["executable"]), _path(script) if script is not None else None))
                    elif kind == "npy":
                        batches.append(NpyDependency(phase_name, identity, _path(workload["processed_directory"])))
        except (KeyError, TypeError, AttributeError) as error:
            raise DependencyError(f"malformed dependency snapshot: {error}") from error
        self._recordings, self._programs, self._batches = tuple(recordings), tuple(programs), tuple(batches)

    @classmethod
    def load(cls, path: str | Path) -> Dependencies:
        """Load an explicit snapshot; failures identify its expected path."""
        try:
            return cls(json.loads(Path(path).read_text(encoding="utf-8")))
        except (OSError, ValueError) as error:
            raise DependencyError(f"cannot load dependency snapshot {path}: {error}") from error

    @classmethod
    def from_env(cls) -> Dependencies:
        """Load WORKFLOW_DEPENDENCIES_PATH from a standard Workflow launch."""
        path = os.environ.get("WORKFLOW_DEPENDENCIES_PATH")
        if not path:
            raise DependencyError("missing WORKFLOW_DEPENDENCIES_PATH; run through Workflow's standard study layout or use Dependencies.load(path)")
        return cls.load(path)

    def raw_json(self) -> object:
        """Return an independent JSON copy, including unknown extensions."""
        return copy.deepcopy(self._raw)

    def recordings(self) -> Selection[RecordingDependency]:
        return Selection(self._recordings)

    def programs(self) -> Selection[ProgramDependency]:
        return Selection(self._programs)

    def npy_batches(self) -> Selection[NpyDependency]:
        return Selection(self._batches)

File: rust/src/task/dependencies.rs
//! Typed, immutable results supplied to a task by its declared dependencies.
//!
//! Selectors preserve snapshot order and never read scientific files. Runtime
//! owns dependency scope; this API cannot discover unrelated output directories.

use serde::Deserialize;
use serde_json::Value;
use std::path::{Path, PathBuf};
use thiserror::Error;

/// Invalid input or a dependency selection with unexpected cardinality.
#[derive(Debug, Error)]
#[non_exhaustive]
pub enum DependencyError {
    /// An environment variable required by the standard launch contract is absent.
    #[error("missing {variable}; run this program through Workflow's standard study layout")]
    MissingEnvironment {
        /// Required variable name.
        variable: &'static str,
    },
    /// A snapshot could not be read.
    #[error("cannot read dependency snapshot `{path}`")]
    Io {
        /// Expected snapshot path.
        path: PathBuf,
        /// Underlying filesystem failure.
        #[source]
        source: std::io::Error,
    },
    /// A known result or the snapshot structure is malformed.
    #[error("invalid dependency snapshot: {0}")]
    Invalid(String),
    /// No result satisfies the selection.
    #[error("no dependency matches {selection}")]
    Missing {
        /// Applied filters.
        selection: String,
    },
    /// Several results satisfy a selection requiring at most one.
    #[error("ambiguous dependency selection {selection}: {matches:?}; select a phase or task")]
    Ambiguous {
        /// Applied filters.
        selection: String,
        /// Candidate source identifiers in snapshot order.
        matches: Vec<String>,
    },
}

/// One validated dependency snapshot. Unknown workload kinds remain in raw JSON.
#[derive(Clone, Debug)]
pub struct Dependencies {
    raw: Value,
    recordings: Vec<RecordingDependency>,
    programs: Vec<ProgramDependency>,
    npy_batches: Vec<NpyDependency>,
}

#[derive(Deserialize)]
struct Phase {
    phase: String,
    tasks: Vec<Task>,
}
#[derive(Deserialize)]
struct Task {
    identity: String,
    output_directory: PathBuf,
    workload: Value,
}
#[derive(Deserialize)]
struct Member {
    identity: String,
    final_iteration: u64,
    output_directory: PathBuf,
}
#[derive(Deserialize)]
struct Unit {
    execution_unit: String,
    members: Vec<Member>,
}
#[derive(Deserialize)]
struct Program {
    executable: PathBuf,
    python_script: Option<PathBuf>,
}
#[derive(Deserialize)]
struct Npy {
    processed_directory: PathBuf,
}

impl Dependencies {
    /// Validates an owned JSON snapshot without opening its referenced files.
    pub fn from_json(raw: Value) -> Result<Self, DependencyError> {
        let phases: Vec<Phase> = decode(raw.clone())?;
        let mut result = Self {
            raw,
            recordings: Vec::new(),
            programs: Vec::new(),
            npy_batches: Vec::new(),
        };
        let mut phases_seen = std::collections::HashSet::new();
        for phase in phases {
            name(&phase.phase)?;
            if !phases_seen.insert(phase.phase.clone()) {
                return Err(DependencyError::Invalid("duplicate phase identity".into()));
            }
            let mut tasks_seen = std::collections::HashSet::new();
            for task in phase.tasks {
                name(&task.identity)?;
                path(&task.output_directory)?;
                if !tasks_seen.insert(task.identity.clone()) {
                    return Err(DependencyError::Invalid(
                        "duplicate task identity in phase".into(),
                    ));
                }
                let source = Source {
                    phase: phase.phase.clone(),
                    task: task.identity,
                };
                let kind = task
                    .workload
                    .get("kind")
                    .and_then(Value::as_str)
                    .ok_or_else(|| {
                        DependencyError::Invalid("workload.kind must be a nonempty string".into())
                    })?;
                name(kind)?;
                match kind {
                    "execution_unit" => {
                        let unit: Unit = decode(task.workload)?;
                        name(&unit.execution_unit)?;
                        let mut members_seen = std::collections::HashSet::new();
                        if unit.members.is_empty() {
                            return Err(DependencyError::Invalid(
                                "execution unit has no members".into(),
                            ));
                        }
                        for member in unit.members {
                            name(&member.identity)?;
                            path(&member.output_directory)?;
                            if !members_seen.insert(member.identity.clone()) {
                                return Err(DependencyError::Invalid(
                                    "duplicate member identity".into(),
                                ));
                            }
                            result.recordings.push(RecordingDependency {
                                source: source.clone(),
                                execution_unit: unit.execution_unit.clone(),
                                member: member.identity,
                                final_iteration: member.final_iteration,
                                directory: member.output_directory,
                            });
                        }
                    }
                    "program" | "python" => {
                        let program: Program = decode(task.workload.clone())?;
                        path(&program.executable)?;
                        if kind == "python" && program.python_script.is_none() {
                            return Err(DependencyError::Invalid(
                                "python workload requires python_script".into(),
                            ));
                        }
                        if let Some(script) = &program.python_script {
                            path(script)?;
                        }
                        result.programs.push(ProgramDependency {
                            source,
                            directory: task.output_directory.join("artifacts"),
                            executable: program.executable,
                            python_script: program.python_script,
                        });
                    }
                    "npy" => {
                        let npy: Npy = decode(task.workload)?;
                        path(&npy.processed_directory)?;
                        result.npy_batches.push(NpyDependency {
                            source,
                            directory: npy.processed_directory,
                        });
                    }
                    _ => {} // Forward-compatible raw access; known kinds are always validated.
                }
            }
        }
        Ok(result)
    }

    /// Loads and validates a standard dependency snapshot from an explicit file.
    pub fn load(path: &Path) -> Result<Self, DependencyError> {
        let bytes = std::fs::read(path).map_err(|source| DependencyError::Io {
            path: path.to_owned(),
            source,
        })?;
        Self::from_json(
            serde_json::from_slice(&bytes)
                .map_err(|error| DependencyError::Invalid(error.to_string()))?,
        )
    }

    /// Loads `WORKFLOW_DEPENDENCIES_PATH`, supplied by a Workflow program launch.
    pub fn from_env() -> Result<Self, DependencyError> {
        let path = std::env::var_os("WORKFLOW_DEPENDENCIES_PATH").ok_or(
            DependencyError::MissingEnvironment {
                variable: "WORKFLOW_DEPENDENCIES_PATH",
            },
        )?;
        Self::load(Path::new(&path))
    }

    /// Borrows the original snapshot, including unknown extension fields/kinds.
    pub const fn raw_json(&self) -> &Value {
        &self.raw
    }
    /// Selects completed member recordings in deterministic snapshot order.
    pub fn recordings(&self) -> Selection<'_, RecordingDependency> {
        Selection::new(&self.recordings)
    }
    /// Selects program and Python task artifact directories.
    pub fn programs(&self) -> Selection<'_, ProgramDependency> {
        Selection::new(&self.programs)
    }
    /// Selects aggregate NPY batches; this does not filter members within a batch.
    pub fn npy_batches(&self) -> Selection<'_, NpyDependency> {
        Selection::new(&self.npy_batches)
    }
}

fn decode<T: serde::de::DeserializeOwned>(value: Value) -> Result<T, DependencyError> {
    serde_json::from_value(value).map_err(|error| DependencyError::Invalid(error.to_string()))
}
fn name(value: &str) -> Result<(), DependencyError> {
    if value.is_empty() || value.trim() != value {
        return Err(DependencyError::Invalid(
            "identifiers must be nonempty without surrounding whitespace".into(),
        ));
    }
    Ok(())
}
fn path(value: &Path) -> Result<(), DependencyError> {
    if !value.is_absolute() {
        return Err(DependencyError::Invalid(format!(
            "expected absolute path, got {}",
            value.display()
        )));
    }
    Ok(())
}

#[derive(Clone, Debug)]
struct Source {
    phase: String,
    task: String,
}

mod sealed {
    pub trait Sealed {}
}
/// Shared identity access for Workflow-owned dependency result types.
///
/// This trait is sealed; downstream implementations are not supported.
pub trait Dependency: sealed::Sealed {
    /// Declared phase key.
    fn phase(&self) -> &str;
    /// Runtime-generated task identity.
    fn task(&self) -> &str;
    /// Human-readable source used in ambiguity diagnostics.
    fn description(&self) -> String;
}

/// Borrowed, filtered dependency results. Filters combine by intersection.
#[derive(Clone, Debug)]
pub struct Selection<'a, T> {
    entries: Vec<&'a T>,
    filters: Vec<String>,
}
impl<'a, T: Dependency> Selection<'a, T> {
    fn new(entries: &'a [T]) -> Self {
        Self {
            entries: entries.iter().collect(),
            filters: Vec::new(),
        }
    }
    fn filter(mut self, label: String, predicate: impl Fn(&T) -> bool) -> Self {
        self.entries.retain(|entry| predicate(entry));
        self.filters.push(label);
        self
    }
    /// Restricts matches to a declared phase key.
    pub fn in_phase(self, phase: &str) -> Self {
        self.filter(format!("phase={phase:?}"), |entry| entry.phase() == phase)
    }
    /// Restricts matches to an exact runtime task identity.
    pub fn task(self, task: &str) -> Self {
        self.filter(format!("task={task:?}"), |entry| entry.task() == task)
    }
    /// Returns exactly one result, reporting missing or ambiguous matches.
    pub fn one(&self) -> Result<&'a T, DependencyError> {
        self.optional()?.ok_or_else(|| DependencyError::Missing {
            selection: self.label(),
        })
    }
    /// Returns zero or one result; several matches produce an ambiguity error.
    pub fn optional(&self) -> Result<Option<&'a T>, DependencyError> {
        match self.entries.as_slice() {
            [] => Ok(None),
            [entry] => Ok(Some(*entry)),
            _ => Err(DependencyError::Ambiguous {
                selection: self.label(),
                matches: self
                    .entries
                    .iter()
                    .map(|entry| entry.description())
                    .collect(),
            }),
        }
    }
    /// Iterates all matching results in snapshot order, including zero matches.
    pub fn iter(&self) -> impl ExactSizeIterator<Item = &'a T> + '_ {
        self.entries.iter().copied()
    }
    fn label(&self) -> String {
        format!(
            "{} [{}]",
            std::any::type_name::<T>(),
            self.filters.join(", ")
        )
    }
}

impl Selection<'_, RecordingDependency> {
    /// Restricts recordings to an exact execution-unit registration key.
    pub fn execution_unit(self, key: &str) -> Self {
        self.filter(format!("execution_unit={key:?}"), |entry| {
            entry.execution_unit == key
        })
    }
    /// Restricts recordings to an exact member identity.
    pub fn member(self, member: &str) -> Self {
        self.filter(format!("member={member:?}"), |entry| entry.member == member)
    }
}

/// A completed member recording.
#[derive(Clone, Debug)]
pub struct RecordingDependency {
    source: Source,
    execution_unit: String,
    member: String,
    final_iteration: u64,
    directory: PathBuf,
}
impl RecordingDependency {
    /// Declared phase key.
    pub fn phase(&self) -> &str {
        &self.source.phase
    }
    /// Runtime-generated task identity.
    pub fn task(&self) -> &str {
        &self.source.task
    }
    /// Execution-unit registration key.
    pub fn execution_unit(&self) -> &str {
        &self.execution_unit
    }
    /// Member identity within its execution unit.
    pub fn member(&self) -> &str {
        &self.member
    }
    /// Final scientific iteration.
    pub fn final_iteration(&self) -> u64 {
        self.final_iteration
    }
    /// Completed recording directory.
    pub fn directory(&self) -> &Path {
        &self.directory
    }
}
impl sealed::Sealed for RecordingDependency {}
impl Dependency for RecordingDependency {
    fn phase(&self) -> &str {
        self.phase()
    }
    fn task(&self) -> &str {
        self.task()
    }
    fn description(&self) -> String {
        format!("{}/{}", self.phase(), self.task()) + &format!("/{}", self.member)
    }
}

/// A completed external executable or Python task.
#[derive(Clone, Debug)]
pub struct ProgramDependency {
    source: Source,
    directory: PathBuf,
    executable: PathBuf,
    python_script: Option<PathBuf>,
}
impl ProgramDependency {
    /// Declared phase key.
    pub fn phase(&self) -> &str {
        &self.source.phase
    }
    /// Runtime-generated task identity.
    pub fn task(&self) -> &str {
        &self.source.task
    }
    /// Standard artifacts directory, excluding logs and snapshots.
    pub fn directory(&self) -> &Path {
        &self.directory
    }
    /// Resolved launcher executable.
    pub fn executable(&self) -> &Path {
        &self.executable
    }
    /// Canonical Python script when this was a Python task.
    pub fn python_script(&self) -> Option<&Path> {
        self.python_script.as_deref()
    }
}
impl sealed::Sealed for ProgramDependency {}
impl Dependency for ProgramDependency {
    fn phase(&self) -> &str {
        self.phase()
    }
    fn task(&self) -> &str {
        self.task()
    }
    fn description(&self) -> String {
        format!("{}/{}", self.phase(), self.task())
    }
}

/// An aggregate converted NPY batch.
#[derive(Clone, Debug)]
pub struct NpyDependency {
    source: Source,
    directory: PathBuf,
}
impl NpyDependency {
    /// Declared phase key.
    pub fn phase(&self) -> &str {
        &self.source.phase
    }
    /// Runtime-generated task identity.
    pub fn task(&self) -> &str {
        &self.source.task
    }
    /// Verified-reader entry directory containing the batch manifest.
    pub fn directory(&self) -> &Path {
        &self.directory
    }
}
impl sealed::Sealed for NpyDependency {}
impl Dependency for NpyDependency {
    fn phase(&self) -> &str {
        self.phase()
    }
    fn task(&self) -> &str {
        self.task()
    }
    fn description(&self) -> String {
        format!("{}/{}", self.phase(), self.task())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    fn snapshot() -> Value {
        serde_json::json!([
            {"phase":"initialize", "tasks":[{"identity":"a", "output_directory":"/run/a", "workload":{"kind":"execution_unit","execution_unit":"init","members":[{"identity":"one","final_iteration":4,"output_directory":"/run/a"}]}}]},
            {"phase":"other", "tasks":[{"identity":"b", "output_directory":"/run/b", "workload":{"kind":"execution_unit","execution_unit":"init","members":[{"identity":"one","final_iteration":5,"output_directory":"/run/b"}]}}]}
        ])
    }
    #[test]
    fn selection_is_explicit_and_preserves_order_and_source() {
        let deps = Dependencies::from_json(snapshot()).unwrap();
        assert!(
            matches!(deps.recordings().execution_unit("init").one(), Err(DependencyError::Ambiguous { matches, .. }) if matches == ["initialize/a/one", "other/b/one"])
        );
        let selected = deps
            .recordings()
            .in_phase("initialize")
            .member("one")
            .one()
            .unwrap();
        assert_eq!(selected.directory(), Path::new("/run/a"));
        assert_eq!(
            deps.recordings()
                .iter()
                .map(|r| r.final_iteration())
                .collect::<Vec<_>>(),
            [4, 5]
        );
        assert!(
            deps.recordings()
                .in_phase("missing")
                .optional()
                .unwrap()
                .is_none()
        );
        assert!(matches!(
            deps.recordings().in_phase("missing").one(),
            Err(DependencyError::Missing { .. })
        ));
        assert_eq!(deps.raw_json(), &snapshot());
    }
    #[test]
    fn malformed_known_results_fail_and_extensions_remain_available() {
        let mut value = snapshot();
        value[0]["tasks"][0]["workload"]["members"][0]["final_iteration"] = "bad".into();
        assert!(Dependencies::from_json(value).is_err());
        let value = serde_json::json!([{"phase":"extension","tasks":[{"identity":"task","output_directory":"/output","workload":{"kind":"future","extra":1}}]}]);
        let deps = Dependencies::from_json(value.clone()).unwrap();
        assert_eq!(deps.raw_json(), &value);
        assert_eq!(deps.recordings().iter().len(), 0);
    }
}
